lib/db/observed.go

```diff
@@ -0,0 +1,265 @@
+func (db *Lowlevel) deleteInvalidPendingFolder(key []byte) error {
+	l.Warnf("Invalid pending folder entry, deleting from database: %x", key)
+	err := db.Delete(key)
+	return err
+}
+
+// Set of devices for which pending folders are allowed, but not specific folder IDs
+type DropListObserved map[protocol.DeviceID]map[string]struct{}
+
+func NewDropListObserved() DropListObserved {
+	return make(DropListObserved)
+}
+
+func (dl DropListObserved) MarkDevice(device protocol.DeviceID) map[string]struct{} {
+	folders, ok := dl[device]
+	if !ok {
+		dl[device] = nil
+	}
+	return folders
+}
+
+func (dl DropListObserved) UnmarkDevice(device protocol.DeviceID) {
+	delete(dl, device)
+}
+
+func (dl DropListObserved) MarkFolder(folder string, devices []protocol.DeviceID) {
+	for _, dev := range devices {
+		folders := dl.MarkDevice(dev)
+		if folders == nil {
+			folders = make(map[string]struct{})
+			dl[dev] = folders
+		}
+		folders[folder] = struct{}{}
+	}
+}
+
+// shouldDropPendingDevice defines how the drop-list is interpreted for pending devices
+func (dl DropListObserved) shouldDropPendingDevice(key []byte, keyer keyer) bool {
+	keyDev := keyer.DeviceFromPendingDeviceKey(key)
+	//FIXME: DeviceIDFromBytes() panics when given a wrong length input.
+	//       It should rather return an error which we'd check for here.
+	if len(keyDev) != protocol.DeviceIDLength {
+		l.Warnf("Invalid pending device entry, deleting from database: %x", key)
+		return true
+	}
+	// Valid entries are looked up in the drop-list, invalid ones cleaned up
+	deviceID := protocol.DeviceIDFromBytes(keyDev)
+	_, dropDevice := dl[deviceID]
+	if dropDevice {
+		l.Debugf("Removing marked pending device %v", deviceID)
+	}
+	return dropDevice
+}
+
+// shouldDropPendingFolder defines how the drop-list is interpreted for pending folders,
+// which is different and more nested than for devices
+func (dl DropListObserved) shouldDropPendingFolder(key []byte, keyer keyer) bool {
+	keyDev, ok := keyer.DeviceFromPendingFolderKey(key)
+	if !ok {
+		l.Warnf("Invalid pending folder entry, deleting from database: %x", key)
+		return true
+	}
+	// Valid entries are looked up in the drop-list, invalid ones cleaned up
+	deviceID := protocol.DeviceIDFromBytes(keyDev)
+	dropFolders, allowDevice := dl[deviceID]
+	// Check the associated set of folders if provided, otherwise drop.
+	if !allowDevice {
+		l.Debugf("Removing pending folder offered by %v", deviceID)
+		return true
+	}
+	if len(dropFolders) == 0 {
+		// Map is empty or nil, skip lookup
+		return false
+	}
+	folderID := keyer.FolderFromPendingFolderKey(key)
+	// Drop only mentioned folder IDs
+	_, dropFolder := dropFolders[string(folderID)]
+	if dropFolder {
+		l.Debugf("Removing marked pending folder %s for %v", folderID, deviceID)
+	}
+	return dropFolder
+}
+
+// CleanPendingDevices removes all pending device entries matching a given set of device IDs
+func (db *Lowlevel) CleanPendingDevices(dropList DropListObserved) {
```

Comment:
This would be cleaner as `CleanPendingDevices(devices []protocol.DeviceID)`

Reply:
True, but then either needs to iterate through the slice for every DB entry (sorting may help) or make a DB lookup for each slice entry (which @imsodin found too expensive).  Making the slice would be an extra step when almost the same data is already collected for `CleanPendingFolders()`.